### ISSUE TITLE
Istio IngressClass and scoped namespaces

### ISSUE THREAD

**Issue by yalctay93** [istio/istio#26919]
Hello, we are using Istio v1.7.0 in our Kubernetes Cluster Version 1.18
Istio has been deployed as "default profile" by using the following yaml 
 ```istioctl profile dump default > profile-default-1.7.0.yaml```

Our currently deployed "IstioOprator" yaml file is shown below as  "**profile-default-1.7.0.yaml**"

Problem: 
- we cannot configure wich ingressClass both Istio-Ingressgateways should use
- we cannot configure which namespaces are in their scope

We are using the Ingress functionality of Istio only because of the cert-manager acme solver procedure, in which the cert-manager will automatically create an ingress-object according to its ```ingressClass``` definition to reach the acme solver pod.

Furthermore, we figured that apparently both Istio-Ingressgateway deployments do not use any ingressClass. Which means it doesn't matter which ingressClass we specify for the cert-manager. It'll still be able to reach its solver pod. So this behaviour brings up the question what is the default ingressClass for the Istio-Ingressgateway? We haven't been able to figure it out in neither the configuration nor inside the created Istio-Ingressgateway pods.

However, we would like to deploy multiple Istio-Ingressgateways in our cluster and seperate these by namespaces. Due to this, we have to be able to configure their ingressClass aswell as define their namespace scopes. Is there any way to configure these with the dumped profile yaml shown below?  Thanks alot!

**cluster-issuer.yaml**
```
apiVersion: cert-manager.io/v1alpha2
kind: ClusterIssuer
metadata:
  name: acme-server
spec:
  acme:
    server: (internal acme server)
    email:
    privateKeySecretRef:
      name: acme-server
    solvers:
    - http01:
        ingress:
          class: istio <-- can be anything, still works
```

 **profile-default-1.7.0.yaml**
```apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-operator-default
spec:
  addonComponents:
    istiocoredns:
      enabled: false
  components:
    base:
      enabled: true
    cni:
      enabled: false
    egressGateways:
    - enabled: false
      k8s:
        env:
        - name: ISTIO_META_ROUTER_MODE
          value: sni-dnat
        hpaSpec:
          maxReplicas: 5
          metrics:
          - resource:
              name: cpu
              targetAverageUtilization: 80
            type: Resource
          minReplicas: 1
          scaleTargetRef:
            apiVersion: apps/v1
            kind: Deployment
            name: istio-egressgateway
        resources:
          limits:
            cpu: 2000m
            memory: 1024Mi
          requests:
            cpu: 100m
            memory: 128Mi
        service:
          ports:
          - name: http2
            port: 80
            targetPort: 8080
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443
        strategy:
          rollingUpdate:
            maxSurge: 100%
            maxUnavailable: 25%
      name: istio-egressgateway
    ingressGateways:
    - name: emea-test-fra-ig-0001
      enabled: true
      label:
        istio: emea-test-fra-ig-0001
      k8s:
        overlays:
          - apiVersion: v1
            kind: Service
            name: emea-test-fra-ig-0001
            patches:
              - path: spec.type
                value: LoadBalancer
              - path: spec.loadBalancerIP
                value: x.x.x.x
              - path: spec.loadBalancerSourceRanges
                value:
                  - x.x.x.x/x
        env:
        - name: ISTIO_META_ROUTER_MODE
          value: sni-dnat
        hpaSpec:
          maxReplicas: 3
          metrics:
          - resource:
              name: cpu
              targetAverageUtilization: 80
            type: Resource
          minReplicas: 1
          scaleTargetRef:
            apiVersion: apps/v1
            kind: Deployment
            name: emea-test-fra-ig-0001
        resources:
          limits:
            cpu: 8000m
            memory: 8192Mi
          requests:
            cpu: 1000m
            memory: 1024Mi
        service:
          ports:
          #- name: status-port
          #  port: 15021
          #  targetPort: 15021
          #  nodePort: 30003
          - name: http
            port: 80
            targetPort: 8080
            nodePort: 30000
          - name: https
            port: 443
            targetPort: 8443
            nodePort: 30001
        strategy:
          rollingUpdate:
            maxSurge: 100%
            maxUnavailable: 25%
    - name: emea-test-fra-ig-0002
      enabled: true
      label:
        istio: emea-test-fra-ig-0002
      k8s:
        overlays:
          - apiVersion: v1
            kind: Service
            name: emea-test-fra-ig-0002
            patches:
              - path: spec.type
                value: LoadBalancer
              - path: spec.loadBalancerIP
                value: x.x.x.x
              - path: spec.loadBalancerSourceRanges
                value:
                  - x.x.x.x/x
        env:
        - name: ISTIO_META_ROUTER_MODE
          value: sni-dnat
        hpaSpec:
          maxReplicas: 3
          metrics:
          - resource:
              name: cpu
              targetAverageUtilization: 80
            type: Resource
          minReplicas: 1
          scaleTargetRef:
            apiVersion: apps/v1
            kind: Deployment
            name: emea-test-fra-ig-0002
        resources:
          limits:
            cpu: 8000m
            memory: 8192Mi
          requests:
            cpu: 1000m
            memory: 1024Mi
        service:
          ports:
          #- name: status-port
          #  port: 15021
          #  targetPort: 15021
          #  nodePort: 30003
          - name: socat
            port: 30000
            targetPort: 2376
            nodePort: 30002
        strategy:
          rollingUpdate:
            maxSurge: 100%
            maxUnavailable: 25%
    istiodRemote:
      enabled: false
    pilot:
      enabled: true
      k8s:
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        readinessProbe:
          httpGet:
            path: /ready
            port: 8080
          initialDelaySeconds: 1
          periodSeconds: 3
          timeoutSeconds: 5
        strategy:
          rollingUpdate:
            maxSurge: 100%
            maxUnavailable: 25%
    policy:
      enabled: false
      k8s:
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        hpaSpec:
          maxReplicas: 5
          metrics:
          - resource:
              name: cpu
              targetAverageUtilization: 80
            type: Resource
          minReplicas: 1
          scaleTargetRef:
            apiVersion: apps/v1
            kind: Deployment
            name: istio-policy
        strategy:
          rollingUpdate:
            maxSurge: 100%
            maxUnavailable: 25%
    telemetry:
      enabled: false
      k8s:
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: GOMAXPROCS
          value: "6"
        hpaSpec:
          maxReplicas: 5
          metrics:
          - resource:
              name: cpu
              targetAverageUtilization: 80
            type: Resource
          minReplicas: 1
          scaleTargetRef:
            apiVersion: apps/v1
            kind: Deployment
            name: istio-telemetry
        replicaCount: 1
        resources:
          limits:
            cpu: 4800m
            memory: 4G
          requests:
            cpu: 1000m
            memory: 1G
        strategy:
          rollingUpdate:
            maxSurge: 100%
            maxUnavailable: 25%
  hub: <internal-hub>
  meshConfig:
    defaultConfig:
      proxyMetadata: {}
    enablePrometheusMerge: true
  profile: default
  tag: 1.7.0
  values:
    base:
      enableCRDTemplates: false
      validationURL: ""
    clusterResources: true
    gateways:
      istio-egressgateway:
        autoscaleEnabled: true
        env: {}
        name: istio-egressgateway
        secretVolumes:
        - mountPath: /etc/istio/egressgateway-certs
          name: egressgateway-certs
          secretName: istio-egressgateway-certs
        - mountPath: /etc/istio/egressgateway-ca-certs
          name: egressgateway-ca-certs
          secretName: istio-egressgateway-ca-certs
        type: ClusterIP
        zvpn: {}
      istio-ingressgateway:
        enabled: false
        applicationPorts: ""
        autoscaleEnabled: true
        debug: info
        domain: ""
        env: {}
        meshExpansionPorts:
        - name: tcp-istiod
          port: 15012
          targetPort: 15012
        - name: tcp-dns-tls
          port: 853
          targetPort: 8853
        name: istio-ingressgateway
        secretVolumes:
        - mountPath: /etc/istio/ingressgateway-certs
          name: ingressgateway-certs
          secretName: istio-ingressgateway-certs
        - mountPath: /etc/istio/ingressgateway-ca-certs
          name: ingressgateway-ca-certs
          secretName: istio-ingressgateway-ca-certs
        type: LoadBalancer
        zvpn: {}
    global:
      arch:
        amd64: 2
        ppc64le: 2
        s390x: 2
      configValidation: true
      controlPlaneSecurityEnabled: true
      defaultNodeSelector: {}
      defaultPodDisruptionBudget:
        enabled: false
      defaultResources:
        requests:
          cpu: 10m
      enableHelmTest: false
      imagePullPolicy: ""
      imagePullSecrets: []
      istioNamespace: istio-system
      istiod:
        enableAnalysis: false
      jwtPolicy: third-party-jwt
      logAsJson: false
      logging:
        level: default:info
      meshExpansion:
        enabled: false
        useILB: false
      meshNetworks: {}
      mountMtlsCerts: false
      multiCluster:
        clusterName: ""
        enabled: false
      network: ""
      omitSidecarInjectorConfigMap: false
      oneNamespace: false
      operatorManageWebhooks: false
      pilotCertProvider: istiod
      priorityClassName: ""
      proxy:
        autoInject: enabled
        clusterDomain: cluster.local
        componentLogLevel: misc:error
        enableCoreDump: false
        excludeIPRanges: ""
        excludeInboundPorts: ""
        excludeOutboundPorts: ""
        image: docker.io_istio_proxyv2
        includeIPRanges: '*'
        logLevel: warning
        privileged: false
        readinessFailureThreshold: 30
        readinessInitialDelaySeconds: 1
        readinessPeriodSeconds: 2
        resources:
          limits:
            cpu: 2000m
            memory: 1024Mi
          requests:
            cpu: 100m
            memory: 128Mi
        statusPort: 15020
        tracer: zipkin
      proxy_init:
        image: docker.io_istio_proxyv2
        resources:
          limits:
            cpu: 2000m
            memory: 1024Mi
          requests:
            cpu: 10m
            memory: 10Mi
      sds:
        token:
          aud: istio-ca
      sts:
        servicePort: 0
      tracer:
        datadog:
          address: $(HOST_IP):8126
        lightstep:
          accessToken: ""
          address: ""
        stackdriver:
          debug: false
          maxNumberOfAnnotations: 200
          maxNumberOfAttributes: 200
          maxNumberOfMessageEvents: 200
        zipkin:
          address: ""
      trustDomain: cluster.local
      useMCP: false
    grafana:
      accessMode: ReadWriteMany
      contextPath: /grafana
      dashboardProviders:
        dashboardproviders.yaml:
          apiVersion: 1
          providers:
          - disableDeletion: false
            folder: istio
            name: istio
            options:
              path: /var/lib/grafana/dashboards/istio
            orgId: 1
            type: file
      datasources:
        datasources.yaml:
          apiVersion: 1
      env: {}
      envSecrets: {}
      image:
        repository: grafana/grafana
        tag: 7.0.5
      nodeSelector: {}
      persist: false
      podAntiAffinityLabelSelector: []
      podAntiAffinityTermLabelSelector: []
      security:
        enabled: false
        passphraseKey: passphrase
        secretName: grafana
        usernameKey: username
      service:
        annotations: {}
        externalPort: 3000
        name: http
        type: ClusterIP
      storageClassName: ""
      tolerations: []
    istiocoredns:
      coreDNSImage: coredns/coredns
      coreDNSPluginImage: istio/coredns-plugin:0.2-istio-1.1
      coreDNSTag: 1.6.2
    istiodRemote:
      injectionURL: ""
    kiali:
      contextPath: /kiali
      createDemoSecret: false
      dashboard:
        auth:
          strategy: login
        grafanaInClusterURL: http://grafana:3000
        jaegerInClusterURL: http://tracing/jaeger
        passphraseKey: passphrase
        secretName: kiali
        usernameKey: username
        viewOnlyMode: false
      hub: quay.io/kiali
      nodeSelector: {}
      podAntiAffinityLabelSelector: []
      podAntiAffinityTermLabelSelector: []
      security:
        cert_file: /kiali-cert/cert-chain.pem
        enabled: false
        private_key_file: /kiali-cert/key.pem
      service:
        annotations: {}
      tag: v1.22
    mixer:
      adapters:
        kubernetesenv:
          enabled: true
        prometheus:
          enabled: true
          metricsExpiryDuration: 10m
        stackdriver:
          auth:
            apiKey: ""
            appCredentials: false
            serviceAccountPath: ""
          enabled: false
          tracer:
            enabled: false
            sampleProbability: 1
        stdio:
          enabled: false
          outputAsJson: false
        useAdapterCRDs: false
      policy:
        adapters:
          kubernetesenv:
            enabled: true
          useAdapterCRDs: false
        autoscaleEnabled: true
        image: mixer
        sessionAffinityEnabled: false
      telemetry:
        autoscaleEnabled: true
        env:
          GOMAXPROCS: "6"
        image: mixer
        loadshedding:
          latencyThreshold: 100ms
          mode: enforce
        nodeSelector: {}
        podAntiAffinityLabelSelector: []
        podAntiAffinityTermLabelSelector: []
        replicaCount: 1
        sessionAffinityEnabled: false
        tolerations: []
    pilot:
      appNamespaces: []
      autoscaleEnabled: true
      autoscaleMax: 5
      autoscaleMin: 1
      configMap: true
      configNamespace: istio-config
      cpu:
        targetAverageUtilization: 80
      enableProtocolSniffingForInbound: true
      enableProtocolSniffingForOutbound: true
      env: {}
      image: docker.io_istio_pilot
      keepaliveMaxServerConnectionAge: 30m
      nodeSelector: {}
      podAntiAffinityLabelSelector: []
      podAntiAffinityTermLabelSelector: []
      policy:
        enabled: false
      replicaCount: 1
      tolerations: []
      traceSampling: 1
    prometheus:
      contextPath: /prometheus
      hub: docker.io/prom
      nodeSelector: {}
      podAntiAffinityLabelSelector: []
      podAntiAffinityTermLabelSelector: []
      provisionPrometheusCert: true
      retention: 6h
      scrapeInterval: 15s
      security:
        enabled: true
      tag: v2.19.2
      tolerations: []
    sidecarInjectorWebhook:
      enableNamespacesByDefault: false
      injectLabel: istio-injection
      objectSelector:
        autoInject: true
        enabled: false
      rewriteAppHTTPProbe: true
    telemetry:
      enabled: true
      v1:
        enabled: false
      v2:
        enabled: true
        metadataExchange:
          wasmEnabled: false
        prometheus:
          enabled: true
          wasmEnabled: false
        stackdriver:
          configOverride: {}
          enabled: false
          logging: false
          monitoring: false
          topology: false
    tracing:
      jaeger:
        accessMode: ReadWriteMany
        hub: docker.io/jaegertracing
        memory:
          max_traces: 50000
        persist: false
        spanStorageType: badger
        storageClassName: ""
        tag: "1.18"
      nodeSelector: {}
      opencensus:
        exporters:
          stackdriver:
            enable_tracing: true
        hub: docker.io/omnition
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: 200m
            memory: 400Mi
        tag: 0.1.9
      podAntiAffinityLabelSelector: []
      podAntiAffinityTermLabelSelector: []
      provider: jaeger
      service:
        annotations: {}
        externalPort: 9411
        name: http-query
        type: ClusterIP
      zipkin:
        hub: docker.io/openzipkin
        javaOptsHeap: 700
        maxSpans: 500000
        node:
          cpus: 2
        probeStartupDelay: 10
        queryPort: 9411
        resources:
          limits:
            cpu: 1000m
            memory: 2048Mi
          requests:
            cpu: 150m
            memory: 900Mi
        tag: 2.20.0
    version: ""

```

**Comment by howardjohn** [comment]
I could not reproduce this. I ran
```
kubectl apply -f - <<EOF
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: istio
  name: ingress
spec:
  rules:
  - host: httpbin.example.com
    http:
      paths:
      - path: /status/*
        backend:
          serviceName: echo
          servicePort: 8000
EOF
```

as well as `kubernetes.io/ingress.class=foo` and no `kubernetes.io/ingress.class` specified. ONLY when kubernetes.io/ingress.class=istio is set did I see any routes configured.

See https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig if you really want to change the ingress class name to be something other than `istio`

**Comment by yalctay93** [comment]
Hello and thank you for the quick response!
...Interesting, so I've tried to rebuild it as well and for us, apparently, it does not matter what ingressClass is defined... 

I wonder what the default value for ingressClass on Istio-Ingressgateway is... shouldn't it always be "istio"? even if it is not specified in meshConfig?

The following Ingress were used to access the fortio echoserver. In both setups, I was able to reach them:
For each setup, I use the following command: ```curl -v http://<hostname>/fortio1```

**1. with ingressClass set to "istio"**

```
kubectl apply -f <<EOF
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: fortio1-ingress
  annotations:
    kubernetes.io/ingress.class: istio
spec:
  rules:
    - host: <hostname>
      http:
        paths:
        - backend:
            serviceName: fortio1-echosrv
            servicePort: 8080
          path: /fortio1
EOF
```

Result: 
```
< HTTP/1.1 200 OK
< content-length: 0
< date: Tue, 01 Sep 2020 07:05:00 GMT
< x-envoy-upstream-service-time: 1
< server: istio-envoy
```

**2. with ingressClass set to "Foo"**
```
kubectl apply -f <<EOF
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: fortio1-ingress
  annotations:
    kubernetes.io/ingress.class: Foo
spec:
  rules:
    - host: <hostname>
      http:
        paths:
        - backend:
            serviceName: fortio1-echosrv
            servicePort: 8080
          path: /fortio1
EOF
```

Result:
```
< HTTP/1.1 200 OK
< content-length: 0
< date: Tue, 01 Sep 2020 07:08:03 GMT
< x-envoy-upstream-service-time: 1
< server: istio-envoy
```

So apparently it does not matter, what ingressClass I set, the Istio-Ingressgateway handles them. 
Even if I set no ingressClass to begin with, it'll still handle the request.
I've set the ingressClass at the following spot in the profile-default.yaml. I am not sure whether that is the correct place to configure it?

```
...
  meshConfig:
    ingressClass: istio
    defaultConfig:
      proxyMetadata: {}
    enablePrometheusMerge: true
  profile: default
  tag: 1.7.0
  values:
    base:
      enableCRDTemplates: false
      validationURL: ""
    clusterResources: true
    gateways:
      istio-egressgateway:
        autoscaleEnabled: true
...
```

**Comment by howardjohn** [comment]
I see, I think the issue is updates. i suspect we filter out the update in https://github.com/istio/istio/blob/658ed6e295bf86535b78f744743f5accba8d5ee9/pilot/pkg/config/kube/ingress/controller.go#L201 which triggers invalid resource.

**Comment by Monkeyanator** [comment]
Should be fixed with #28336, going to go ahead and close
